File: frontend/src/services/stockProductService.ts
import { apiClient } from './apiClient'

export enum StatusEnum {
  Ativo = 1,
  Vendido = 2,
  Expirado = 3,
}

export interface StockProductResponse {
  id: number
  internalCode: string
  productName: string
  section: string
  expirationDate: string
  unitType: string
  originalPrice: number
  promotionalPrice: number
  costPrice: number
  priority: number
  status: StatusEnum
  statusDescription: string
  updatedBy: string
  controlDate: string
  quantity: number
}

export interface CreateStockProductDto {
  internalCode: string
  expirationDate: string
  unitType: string
  originalPrice: number
  promotionalPrice: number
  costPrice: number
  priority: number
  status: StatusEnum
  updatedBy: string
  controlDate: string
}

export interface UpdateStockProductDto {
  expirationDate?: string
  unitType?: string
  originalPrice?: number
  promotionalPrice?: number
  costPrice?: number
  priority?: number
  status?: StatusEnum
  updatedBy?: string
  controlDate?: string
}

class StockProductService {
  private basePath = '/stock-products'

  async getAll(): Promise<StockProductResponse[]> {
    const response = await apiClient.get<StockProductResponse[]>(this.basePath)
    return response.data
  }

  async getById(id: number): Promise<StockProductResponse> {
    const response = await apiClient.get<StockProductResponse>(`${this.basePath}/${id}`)
    return response.data
  }

  async getByInternalCode(internalCode: string): Promise<StockProductResponse[]> {
    const response = await apiClient.get<StockProductResponse[]>(
      `${this.basePath}/by-code/${internalCode}`
    )
    return response.data
  }

  async getByStatus(status: StatusEnum): Promise<StockProductResponse[]> {
    const response = await apiClient.get<StockProductResponse[]>(
      `${this.basePath}/by-status/${status}`
    )
    return response.data
  }

  async getExpiringProducts(expirationDate?: string): Promise<StockProductResponse[]> {
    const params = expirationDate ? `?expirationDate=${expirationDate}` : ''
    const response = await apiClient.get<StockProductResponse[]>(
      `${this.basePath}/expiring${params}`
    )
    return response.data
  }

  async create(dto: CreateStockProductDto): Promise<StockProductResponse> {
    const response = await apiClient.post<StockProductResponse>(this.basePath, dto)
    return response.data
  }

  async update(id: number, dto: UpdateStockProductDto): Promise<StockProductResponse> {
    const response = await apiClient.put<StockProductResponse>(
      `${this.basePath}/${id}`,
      dto
    )
    return response.data
  }

  async delete(id: number): Promise<void> {
    await apiClient.delete(`${this.basePath}/${id}`)
  }

  async exists(id: number): Promise<boolean> {
    try {
      await apiClient.get(`${this.basePath}/${id}`)
      return true
    } catch (error: any) {
      if (error.response?.status === 404) {
        return false
      }
      throw error
    }
  }
}

export const stockProductService = new StockProductService()

File: API/ApiVilaTrack/ApiVilaTrack/Services/StockProductService.cs
using ApiVilaTrack.Data;
using ApiVilaTrack.Dtos;
using ApiVilaTrack.Entities;
using ApiVilaTrack.Repositories;
using Microsoft.EntityFrameworkCore;

namespace ApiVilaTrack.Services;

public class StockProductService
{
    private readonly StockProductRepository _repository;
    private readonly AppDbContext _context;

    public StockProductService(StockProductRepository repository, AppDbContext context)
    {
        _repository = repository;
        _context = context;
    }

    public async Task<IEnumerable<StockProductResponseDto>> GetAllAsync()
    {
        var stockProducts = await _repository.GetAllAsync();
        return stockProducts.Select(MapToResponseDto);
    }

    public async Task<StockProductResponseDto?> GetByIdAsync(int id)
    {
        var stockProduct = await _repository.GetByIdAsync(id);
        return stockProduct != null ? MapToResponseDto(stockProduct) : null;
    }

    public async Task<IEnumerable<StockProductResponseDto>> GetByInternalCodeAsync(string internalCode)
    {
        var stockProducts = await _repository.GetByInternalCodeAsync(internalCode);
        return stockProducts.Select(MapToResponseDto);
    }

    public async Task<IEnumerable<StockProductResponseDto>> GetByStatusAsync(StatusEnum status)
    {
        var stockProducts = await _repository.GetByStatusAsync(status);
        return stockProducts.Select(MapToResponseDto);
    }

    public async Task<IEnumerable<StockProductResponseDto>> GetExpiringProductsAsync(DateOnly expirationDate)
    {
        var stockProducts = await _repository.GetExpiringProductsAsync(expirationDate);
        return stockProducts.Select(MapToResponseDto);
    }

    public async Task<StockProductResponseDto> CreateAsync(CreateStockProductDto dto)
    {
        // Verificar se o produto existe no catálogo
        var catalogExists = await _context.Catalogs.AnyAsync(c => c.InternalCode == dto.InternalCode);
        if (!catalogExists)
            throw new ArgumentException($"Produto com código '{dto.InternalCode}' não encontrado no catálogo.");

        var stockProduct = new StockProduct
        {
            InternalCode = dto.InternalCode,
            ExpirationDate = dto.ExpirationDate,
            UnitType = dto.UnitType,
            OriginalPrice = dto.OriginalPrice,
            PromotionalPrice = dto.PromotionalPrice,
            CostPrice = dto.CostPrice,
            Priority = dto.Priority,
            Status = dto.Status,
            UpdatedBy = dto.UpdatedBy,
            ControlDate = dto.ControlDate
        };

        var created = await _repository.AddAsync(stockProduct);
        return MapToResponseDto(created);
    }

    public async Task<StockProductResponseDto?> UpdateAsync(int id, UpdateStockProductDto dto)
    {
        var existingProduct = await _repository.GetByIdAsync(id);
        if (existingProduct == null)
            return null;

        // Atualizar apenas os campos fornecidos
        if (dto.ExpirationDate.HasValue)
            existingProduct.ExpirationDate = dto.ExpirationDate.Value;
        
        if (!string.IsNullOrEmpty(dto.UnitType))
            existingProduct.UnitType = dto.UnitType;
        
        if (dto.OriginalPrice.HasValue)
            existingProduct.OriginalPrice = dto.OriginalPrice.Value;
        
        if (dto.PromotionalPrice.HasValue)
            existingProduct.PromotionalPrice = dto.PromotionalPrice.Value;
        
        if (dto.CostPrice.HasValue)
            existingProduct.CostPrice = dto.CostPrice.Value;
        
        if (dto.Priority.HasValue)
            existingProduct.Priority = dto.Priority.Value;
        
        if (dto.Status.HasValue)
            existingProduct.Status = dto.Status.Value;
        
        if (!string.IsNullOrEmpty(dto.UpdatedBy))
            existingProduct.UpdatedBy = dto.UpdatedBy;
        
        if (dto.ControlDate.HasValue)
            existingProduct.ControlDate = dto.ControlDate.Value;

        var updated = await _repository.UpdateAsync(id, existingProduct);
        if (!updated)
            return null;

        var updatedProduct = await _repository.GetByIdAsync(id);
        return updatedProduct != null ? MapToResponseDto(updatedProduct) : null;
    }

    public async Task<bool> DeleteAsync(int id)
    {
        return await _repository.DeleteAsync(id);
    }

    public async Task<bool> ExistsAsync(int id)
    {
        return await _repository.ExistsAsync(id);
    }

    private StockProductResponseDto MapToResponseDto(StockProduct stockProduct)
    {
        return new StockProductResponseDto(
            stockProduct.Id,
            stockProduct.InternalCode,
            stockProduct.Catalog?.Name ?? "Produto não encontrado",
            stockProduct.Catalog?.Section ?? "Seção não definida",
            stockProduct.ExpirationDate,
            stockProduct.UnitType,
            stockProduct.OriginalPrice,
            stockProduct.PromotionalPrice,
            stockProduct.CostPrice,
            stockProduct.Priority,
            stockProduct.Status,
            stockProduct.StatusEntity?.Description ?? GetStatusDescription(stockProduct.Status),
            stockProduct.UpdatedBy,
            stockProduct.ControlDate,
            stockProduct.Catalog?.Quantity ?? 0
        );
    }

    /// <summary>
    /// Retorna a descrição textual do status do produto.
    /// </summary>
    private string GetStatusDescription(StatusEnum status)
    {
        return status switch
        {
            StatusEnum.Ativo => "Ativo",
            StatusEnum.Vendido => "Vendido",
            StatusEnum.Expirado => "Expirado",
            _ => "Desconhecido"
        };
    }
}